File: backend/src/main/java/dev/arc/engine/Expressions.java
package dev.arc.engine;

import dev.arc.error.ArcException;
import java.math.BigDecimal;
import java.math.MathContext;
import java.util.*;
import java.util.regex.Pattern;

/**
 * A bounded expression language with decimal arithmetic and a function registry. No reflection,
 * scripting, IO, or arbitrary calls.
 */
public final class Expressions {
  public static final MathContext MATH = MathContext.DECIMAL128;
  private static final Pattern TOKEN =
      Pattern.compile(
          "\\s*(?:(\\d+(?:\\.\\d+)?(?:[eE][+-]?\\d+)?|\\.\\d+)|([A-Za-z_][A-Za-z_0-9.]*)|(\"(?:[^\"\\\\]|\\\\.)*\"|'(?:[^'\\\\]|\\\\.)*')|(&&|\\|\\||==|!=|<>|<=|>=|[=^\\[\\]+*/%<>()!,\\-]))");

  private Expressions() {}

  public interface Expr {
    Object eval(Map<String, Object> variables);
  }

  public record Compiled(Expr expr, Set<String> variables) {
    public Object evaluate(Map<String, Object> scope) {
      WORK.set(0);
      try {
        return bounded(expr.eval(scope));
      } finally {
        WORK.remove();
      }
    }
  }

  public static Compiled compile(String source) {
    if (source == null || source.isBlank()) throw ArcException.invalid("Expression is required");
    if (source.length() > 2000) throw ArcException.invalid("Expression exceeds 2,000 characters");
    var parser = new Parser(source.trim());
    Expr expr = parser.parse(0);
    if (!parser.peek().equals("<end>"))
      throw ArcException.invalid("Unexpected token: " + parser.peek());
    return new Compiled(expr, Set.copyOf(parser.variables));
  }

  public static Object evaluate(String source, Map<String, Object> scope) {
    return compile(source).evaluate(scope);
  }

  public static BigDecimal number(Object value) {
    if (!(value instanceof Number))
      throw ArcException.invalid("Expected a number, got " + type(value));
    try {
      return (BigDecimal)
          bounded(value instanceof BigDecimal b ? b : new BigDecimal(value.toString()));
    } catch (NumberFormatException e) {
      throw ArcException.invalid("Invalid numeric value");
    }
  }

  public static boolean bool(Object value) {
    if (!(value instanceof Boolean b))
      throw ArcException.invalid("Expected a boolean, got " + type(value));
    return b;
  }

  private static final ThreadLocal<Integer> WORK = ThreadLocal.withInitial(() -> 0);

  private static void tick() {
    int n = WORK.get() + 1;
    if (n > 10000) throw ArcException.invalid("Expression exceeds 10,000 operations");
    WORK.set(n);
  }

  public static Object bounded(Object value) {
    bound(value, 0, new int[] {0});
    return value;
  }

  private static void bound(Object value, int depth, int[] count) {
    if (depth > 8 || ++count[0] > 10000)
      throw ArcException.invalid("Value exceeds collection depth or size limit");
    if (value instanceof BigDecimal n && (n.precision() > 100 || Math.abs((long) n.scale()) > 100))
      throw ArcException.invalid("Number exceeds supported precision or magnitude");
    if (value instanceof Number n && !Double.isFinite(n.doubleValue()))
      throw ArcException.invalid("Number must be finite");
    if (value instanceof String s && s.length() > 2000)
      throw ArcException.invalid("String exceeds 2,000 characters");
    if (value instanceof List<?> xs) {
      if (xs.size() > 1000) throw ArcException.invalid("Array exceeds 1,000 items");
      for (Object x : xs) bound(x, depth + 1, count);
    }
    if (value instanceof Map<?, ?> m) {
      if (m.size() > 1000) throw ArcException.invalid("Object exceeds 1,000 fields");
      for (var e : m.entrySet()) {
        bound(e.getKey(), depth + 1, count);
        bound(e.getValue(), depth + 1, count);
      }
    }
  }

  public static boolean equal(Object a, Object b) {
    return a instanceof Number && b instanceof Number
        ? number(a).compareTo(number(b)) == 0
        : Objects.equals(a, b);
  }

  private static String type(Object o) {
    return o == null ? "null" : o.getClass().getSimpleName();
  }

  private static Object binary(String op, Object a, Object b) {
    if (Set.of("==", "!=", "=", "<>").contains(op)) {
      boolean same =
          a instanceof Number && b instanceof Number
              ? number(a).compareTo(number(b)) == 0
              : Objects.equals(a, b);
      return (op.equals("==") || op.equals("=")) == same;
    }
    if (Set.of("<", "<=", ">", ">=").contains(op)) {
      int cmp;
      if (a instanceof String x && b instanceof String y) cmp = x.compareTo(y);
      else cmp = number(a).compareTo(number(b));
      return switch (op) {
        case "<" -> cmp < 0;
        case "<=" -> cmp <= 0;
        case ">" -> cmp > 0;
        default -> cmp >= 0;
      };
    }
    BigDecimal x = number(a), y = number(b);
    if ((op.equals("/") || op.equals("%")) && y.signum() == 0)
      throw ArcException.invalid("Division by zero");
    return bounded(
        switch (op) {
          case "+" -> x.add(y, MATH);
          case "-" -> x.subtract(y, MATH);
          case "*" -> x.multiply(y, MATH);
          case "/" -> x.divide(y, MATH);
          case "%" -> x.remainder(y, MATH);
          case "^" -> {
            int exponent;
            try {
              exponent = y.intValueExact();
            } catch (ArithmeticException e) {
              throw ArcException.invalid("Exponent must be an integer");
            }
            if (Math.abs((long) exponent) > 100)
              throw ArcException.invalid("Exponent must be -100 to 100");
            try {
              yield exponent < 0
                  ? BigDecimal.ONE.divide(x.pow(-exponent, MATH), MATH)
                  : x.pow(exponent, MATH);
            } catch (ArithmeticException e) {
              throw ArcException.invalid("Invalid power");
            }
          }
          default -> throw ArcException.invalid("Unknown operator: " + op);
        });
  }

  private static Object function(String name, List<Expr> args, Map<String, Object> scope) {
    tick();
    if (name.equals("COALESCE")) {
      for (Expr argument : args) {
        Object value = argument.eval(scope);
        if (value != null) return value;
      }
      return null;
    }
    if (name.equals("IF")) return args.get(bool(args.get(0).eval(scope)) ? 1 : 2).eval(scope);
    if (Set.of("ISERROR", "ISERR", "ISNA").contains(name)) {
      try {
        args.getFirst().eval(scope);
        return false;
      } catch (ArcException e) {
        boolean na = e.getMessage().contains("#N/A");
        return name.equals("ISERROR") || name.equals("ISNA") && na || name.equals("ISERR") && !na;
      }
    }
    if (name.equals("IFERROR")) {
      try {
        return args.get(0).eval(scope);
      } catch (ArcException e) {
        return args.get(1).eval(scope);
      }
    }
    if (name.equals("AND")) return args.stream().allMatch(a -> bool(a.eval(scope)));
    if (name.equals("OR")) return args.stream().anyMatch(a -> bool(a.eval(scope)));
    if (name.equals("SWITCH")) {
      Object v = args.getFirst().eval(scope);
      for (int i = 1; i + 1 < args.size(); i += 2)
        if (equal(v, args.get(i).eval(scope))) return args.get(i + 1).eval(scope);
      if (args.size() % 2 == 0) return args.getLast().eval(scope);
      throw ArcException.invalid("SWITCH has no matching case or default");
    }
    return bounded(Functions.call(name, args.stream().map(a -> a.eval(scope)).toList()));
  }

  private static int priority(String op) {
    return switch (op) {
      case "||", "OR" -> 1;
      case "&&", "AND" -> 2;
      case "==", "!=", "=", "<>" -> 3;
      case "<", "<=", ">", ">=" -> 4;
      case "+", "-" -> 5;
      case "*", "/", "%" -> 6;
      case "^" -> 8;
      default -> -1;
    };
  }

  private static final class Parser {
    private final List<String> tokens = new ArrayList<>();
    private final Set<String> variables = new HashSet<>();
    private int index, depth;
    private final Set<String> locals = new HashSet<>();

    Parser(String source) {
      var matcher = TOKEN.matcher(source);
      int position = 0;
      while (position < source.length()) {
        matcher.region(position, source.length());
        if (!matcher.lookingAt())
          throw ArcException.invalid("Invalid expression near character " + (position + 1));
        tokens.add(matcher.group().trim());
        position = matcher.end();
        if (tokens.size() > 256) throw ArcException.invalid("Expression exceeds 256 tokens");
      }
      tokens.add("<end>");
    }

    String peek() {
      return tokens.get(index);
    }

    String take() {
      return tokens.get(index++);
    }

    void expect(String token) {
      if (!peek().equals(token))
        throw ArcException.invalid("Expected '" + token + "', got '" + peek() + "'");
      take();
    }

    Expr parse(int min) {
      if (++depth > 48) throw ArcException.invalid("Expression nesting exceeds 48 levels");
      Expr left = atom();
      while (priority(peek()) >= min) {
        String op = take();
        Expr a = left, b = parse(priority(op) + (op.equals("^") ? 0 : 1));
        left =
            switch (op) {
              case "&&", "AND" -> s -> bool(a.eval(s)) && bool(b.eval(s));
              case "||", "OR" -> s -> bool(a.eval(s)) || bool(b.eval(s));
              default -> s -> binary(op, a.eval(s), b.eval(s));
            };
      }
      depth--;
      return left;
    }

    Expr atom() {
      String token = take();
      if (token.equals("<end>")) throw ArcException.invalid("Incomplete expression");
      if (token.equals("[")) {
        var items = new ArrayList<Expr>();
        if (!peek().equals("]")) {
          items.add(parse(0));
          while (peek().equals(",")) {
            take();
            items.add(parse(0));
          }
        }
        expect("]");
        return s -> bounded(items.stream().map(x -> x.eval(s)).toList());
      }
      if (token.equals("(")) {
        Expr nested = parse(0);
        expect(")");
        return nested;
      }
      if (Set.of("!", "-", "+").contains(token)) {
        Expr operand = parse(7);
        return s ->
            token.equals("!")
                ? !bool(operand.eval(s))
                : token.equals("-")
                    ? number(operand.eval(s)).negate(MATH)
                    : number(operand.eval(s));
      }
      if (token.startsWith("\"") || token.startsWith("'")) {
        StringBuilder value = new StringBuilder();
        for (int i = 1; i < token.length() - 1; i++) {
          char c = token.charAt(i);
          if (c == '\\') {
            c = token.charAt(++i);
            c =
                switch (c) {
                  case 'n' -> '\n';
                  case 't' -> '\t';
                  case 'r' -> '\r';
                  default -> c;
                };
          }
          value.append(c);
        }
        return s -> value.toString();
      }
      if (Character.isDigit(token.charAt(0)) || token.charAt(0) == '.') {
        try {
          BigDecimal n = number(new BigDecimal(token));
          return s -> n;
        } catch (NumberFormatException e) {
          throw ArcException.invalid("Invalid numeric literal");
        }
      }
      if (token.equalsIgnoreCase("true") && !peek().equals("(")) return s -> true;
      if (token.equalsIgnoreCase("false") && !peek().equals("(")) return s -> false;
      if (token.equalsIgnoreCase("null")) return s -> null;
      if (!token.matches("[A-Za-z_][A-Za-z_0-9.]*"))
        throw ArcException.invalid("Unexpected token: " + token);
      if (peek().equals("(")) {
        take();
        String name = token.toUpperCase(Locale.ROOT);
        if (Set.of("MAP", "FILTER", "ALL", "ANY", "REDUCE").contains(name)) {
          Expr collection = parse(0);
          expect(",");
          String local = take();
          if (!Validator.identifier(local))
            throw ArcException.invalid("Collection function needs a local item identifier");
          expect(",");
          String accumulator = null;
          Expr initial = null;
          if (name.equals("REDUCE")) {
            accumulator = take();
            if (!Validator.identifier(accumulator) || local.equals(accumulator))
              throw ArcException.invalid("REDUCE needs distinct item and accumulator identifiers");
            expect(",");
            initial = parse(0);
            expect(",");
          }
          var before = new HashSet<>(locals);
          locals.add(local);
          if (accumulator != null) locals.add(accumulator);
          Expr body = parse(0);
          locals.clear();
          locals.addAll(before);
          expect(")");
          String acc = accumulator;
          Expr init = initial;
          return s -> {
            var xs = Functions.array(collection.eval(s));
            var result = new ArrayList<Object>();
            Object total = init == null ? null : init.eval(s);
            for (Object item : xs) {
              tick();
              var child = new HashMap<>(s);
              child.put(local, item);
              if (acc != null) child.put(acc, total);
              Object v = body.eval(child);
              switch (name) {
                case "MAP" -> result.add(v);
                case "FILTER" -> {
                  if (bool(v)) result.add(item);
                }
                case "ALL" -> {
                  if (!bool(v)) return false;
                }
                case "ANY" -> {
                  if (bool(v)) return true;
                }
                case "REDUCE" -> total = v;
              }
            }
            return bounded(
                name.equals("REDUCE")
                    ? total
                    : name.equals("ALL") ? true : name.equals("ANY") ? false : result);
          };
        }
        var args = new ArrayList<Expr>();
        if (!peek().equals(")")) {
          args.add(parse(0));
          while (peek().equals(",")) {
            take();
            args.add(parse(0));
          }
        }
        expect(")");
        Functions.arity(name, args.size());
        return s -> function(name, args, s);
      }
      String root = token.split("\\.")[0];
      if (!locals.contains(root)) variables.add(root);
      return s -> {
        if (!s.containsKey(root)) throw ArcException.invalid("Unknown variable: " + root);
        return token.equals(root)
            ? s.get(root)
            : Functions.get(s.get(root), token.substring(root.length() + 1), null);
      };
    }
  }
}

File: backend-rails/lib/arc/expressions.rb
require "set"
require "bigdecimal"

module Arc
  module Expressions
    TOKEN = /\G\s*(?:(\d+(?:\.\d+)?(?:[eE][+-]?\d+)?|\.\d+)|([A-Za-z_][A-Za-z_0-9.]*)|("(?:[^"\\]|\\.)*"|'(?:[^'\\]|\\.)*')|(&&|\|\||==|!=|<>|<=|>=|[=^\[\]+*\/%<>()!,\-]))/
    PRIORITY = { "||" => 1, "OR" => 1, "&&" => 2, "AND" => 2, "==" => 3, "!=" => 3, "=" => 3, "<>" => 3,
                 "<" => 4, "<=" => 4, ">" => 4, ">=" => 4, "+" => 5, "-" => 5, "*" => 6, "/" => 6, "%" => 6, "^" => 8 }.freeze
    module_function

    def invalid(message)
      raise ArcError.new(422, message)
    end

    def compile(source)
      invalid("Expression is required") if source.nil? || source.strip.empty?
      invalid("Expression exceeds 2,000 characters") if source.length > 2000
      parser = Parser.new(source.strip)
      ast = parser.parse(0)
      invalid("Unexpected token: #{parser.peek}") unless parser.peek == "<end>"
      Compiled.new(ast, parser.variables)
    end

    def evaluate(source, scope)
      compile(source).evaluate(scope)
    end

    def type(value)
      case value
      when nil then "null"
      when String then "String"
      when true, false then "Boolean"
      when BigDecimal then "BigDecimal"
      when Integer then "Integer"
      when Float then "Double"
      when Array then "ArrayList"
      when Hash then "LinkedHashMap"
      else value.class.name
      end
    end

    def number(value)
      invalid("Expected a number, got #{type(value)}") unless value.is_a?(Numeric)
      bounded(value.is_a?(BigDecimal) ? value : BigDecimal(value.to_s))
    rescue ArgumentError
      invalid("Invalid numeric value")
    end

    def bool(value)
      invalid("Expected a boolean, got #{type(value)}") unless value == true || value == false
      value
    end

    def bounded(value, depth = 0, count = [0])
      count[0] += 1
      invalid("Value exceeds collection depth or size limit") if depth > 8 || count[0] > 10_000
      if value.is_a?(BigDecimal) && value.finite?
        digits = value.split[1].length
        invalid("Number exceeds supported precision or magnitude") if digits > 100 || (digits - value.exponent).abs > 100
      end
      invalid("Number must be finite") if value.is_a?(Numeric) && !value.to_f.finite?
      invalid("String exceeds 2,000 characters") if value.is_a?(String) && value.encode("UTF-16BE").bytesize > 4000
      if value.is_a?(Array)
        invalid("Array exceeds 1,000 items") if value.size > 1000
        value.each { |v| bounded(v, depth + 1, count) }
      elsif value.is_a?(Hash)
        invalid("Object exceeds 1,000 fields") if value.size > 1000
        value.each { |k, v| bounded(k, depth + 1, count); bounded(v, depth + 1, count) }
      end
      value
    end

    def equal(a, b)
      a.is_a?(Numeric) && b.is_a?(Numeric) ? number(a) == number(b) : a == b
    end

    def decimal(value)
      value.round(34 - value.exponent, BigDecimal::ROUND_HALF_EVEN)
    end

    def binary(op, a, b)
      return (%w[== =].include?(op)) == equal(a, b) if %w[== != = <>].include?(op)
      if %w[< <= > >=].include?(op)
        cmp = a.is_a?(String) && b.is_a?(String) ? a <=> b : number(a) <=> number(b)
        return { "<" => cmp < 0, "<=" => cmp <= 0, ">" => cmp > 0, ">=" => cmp >= 0 }.fetch(op)
      end
      x, y = number(a), number(b)
      invalid("Division by zero") if %w[/ %].include?(op) && y.zero?
      result = case op
      when "+" then x.add(y, 34)
      when "-" then x.sub(y, 34)
      when "*" then x.mult(y, 34)
      when "/" then x.div(y, 34)
      when "%" then decimal(x.remainder(y))
      when "^"
        invalid("Exponent must be an integer") unless y.frac.zero?
        invalid("Exponent must be -100 to 100") if y.abs > 100
        invalid("Invalid power") if x.zero? && y.negative?
        y.negative? ? BigDecimal("1").div(x.power(-y.to_i, 34), 34) : x.power(y.to_i, 34)
      else invalid("Unknown operator: #{op}")
      end
      bounded(result)
    end

    class Compiled
      attr_reader :variables
      def initialize(ast, variables)
        @ast, @variables = ast, variables.freeze
      end

      def evaluate(scope)
        BigDecimal.save_rounding_mode do
          BigDecimal.mode(BigDecimal::ROUND_MODE, BigDecimal::ROUND_HALF_EVEN)
          Expressions.bounded(Evaluation.new.run(@ast, scope))
        end
      end
    end

    class Evaluation
      def initialize
        @work = 0
      end

      def tick
        @work += 1
        Expressions.invalid("Expression exceeds 10,000 operations") if @work > 10_000
      end

      def run(ast, scope)
        kind, *args = ast
        case kind
        when :literal then args[0]
        when :variable
          root, path = args
          Expressions.invalid("Unknown variable: #{root}") unless scope.key?(root)
          path ? Functions.get(scope[root], path, nil) : scope[root]
        when :array then Expressions.bounded(args[0].map { |a| run(a, scope) })
        when :unary
          op, expr = args
          value = run(expr, scope)
          op == "!" ? !Expressions.bool(value) : op == "-" ? Expressions.decimal(-Expressions.number(value)) : Expressions.number(value)
        when :binary
          op, a, b = args
          return Expressions.bool(run(a, scope)) && Expressions.bool(run(b, scope)) if %w[&& AND].include?(op)
          return Expressions.bool(run(a, scope)) || Expressions.bool(run(b, scope)) if %w[|| OR].include?(op)
          Expressions.binary(op, run(a, scope), run(b, scope))
        when :function then function(args[0], args[1], scope)
        when :collection
          name, collection, local, accumulator, initial, body = args
          xs = Functions.array(run(collection, scope))
          total = initial && run(initial, scope)
          result = []
          xs.each do |item|
            tick
            child = scope.merge(local => item)
            child[accumulator] = total if accumulator
            value = run(body, child)
            case name
            when "MAP" then result << value
            when "FILTER" then result << item if Expressions.bool(value)
            when "ALL" then return false unless Expressions.bool(value)
            when "ANY" then return true if Expressions.bool(value)
            when "REDUCE" then total = value
            end
          end
          Expressions.bounded(name == "REDUCE" ? total : name == "ALL" ? true : name == "ANY" ? false : result)
        end
      end

      def function(name, args, scope)
        tick
        return run(args[Expressions.bool(run(args[0], scope)) ? 1 : 2], scope) if name == "IF"
        if %w[ISERROR ISERR ISNA].include?(name)
          begin
            run(args[0], scope)
            return false
          rescue ArcError => e
            na = e.message.include?("#N/A")
            return name == "ISERROR" || (name == "ISNA" && na) || (name == "ISERR" && !na)
          end
        end
        if name == "IFERROR"
          begin
            return run(args[0], scope)
          rescue ArcError
            return run(args[1], scope)
          end
        end
        return args.all? { |a| Expressions.bool(run(a, scope)) } if name == "AND"
        return args.any? { |a| Expressions.bool(run(a, scope)) } if name == "OR"
        if name == "SWITCH"
          value = run(args[0], scope)
          (1...args.size - 1).step(2) { |i| return run(args[i + 1], scope) if Expressions.equal(value, run(args[i], scope)) }
          return run(args[-1], scope) if args.size.even?
          Expressions.invalid("SWITCH has no matching case or default")
        end
        Expressions.bounded(Functions.call(name, args.map { |a| run(a, scope) }))
      end
    end

    class Parser
      attr_reader :variables
      def initialize(source)
        @tokens, @variables, @locals = [], Set.new, Set.new
        @index = @depth = 0
        position = 0
        while position < source.length
          match = TOKEN.match(source, position)
          Expressions.invalid("Invalid expression near character #{position + 1}") unless match
          @tokens << match[0].strip
          position = match.end(0)
          Expressions.invalid("Expression exceeds 256 tokens") if @tokens.size > 256
        end
        @tokens << "<end>"
      end

      def peek
        @tokens[@index] || "<end>"
      end

      def take
        value = peek
        @index += 1
        value
      end

      def expect(token)
        Expressions.invalid("Expected '#{token}', got '#{peek}'") unless peek == token
        take
      end

      def parse(min)
        @depth += 1
        Expressions.invalid("Expression nesting exceeds 48 levels") if @depth > 48
        left = atom
        while PRIORITY.fetch(peek, -1) >= min
          op = take
          left = [:binary, op, left, parse(PRIORITY.fetch(op) + (op == "^" ? 0 : 1))]
        end
        @depth -= 1
        left
      end

      def atom
        token = take
        Expressions.invalid("Incomplete expression") if token == "<end>"
        if token == "["
          values = arguments("]")
          return [:array, values]
        end
        if token == "("
          nested = parse(0)
          expect(")")
          return nested
        end
        return [:unary, token, parse(7)] if %w[! - +].include?(token)
        if token.start_with?('"', "'")
          return [:literal, token[1...-1].gsub(/\\(.)/m) { { "n" => "\n", "t" => "\t", "r" => "\r" }.fetch(Regexp.last_match(1), Regexp.last_match(1)) }]
        end
        return [:literal, Expressions.number(BigDecimal(token))] if /\A[\d.]/.match?(token)
        return [:literal, token.casecmp?("true")] if %w[true false].include?(token.downcase) && peek != "("
        return [:literal, nil] if token.casecmp?("null")
        Expressions.invalid("Unexpected token: #{token}") unless /\A[A-Za-z_][A-Za-z_0-9.]*\z/.match?(token)
        if peek == "("
          take
          name = token.upcase
          if %w[MAP FILTER ALL ANY REDUCE].include?(name)
            collection = parse(0)
            expect(",")
            local = take
            Expressions.invalid("Collection function needs a local item identifier") unless Validator.identifier?(local)
            expect(",")
            accumulator = initial = nil
            if name == "REDUCE"
              accumulator = take
              Expressions.invalid("REDUCE needs distinct item and accumulator identifiers") unless Validator.identifier?(accumulator) && accumulator != local
              expect(",")
              initial = parse(0)
              expect(",")
            end
            before = @locals.dup
            @locals << local
            @locals << accumulator if accumulator
            body = parse(0)
            @locals = before
            expect(")")
            return [:collection, name, collection, local, accumulator, initial, body]
          end
          args = arguments(")")
          Functions.arity(name, args.size)
          return [:function, name, args]
        end
        root, path = token.split(".", 2)
        @variables << root unless @locals.include?(root)
        [:variable, root, path]
      end

      def arguments(close)
        args = []
        unless peek == close
          args << parse(0)
          while peek == ","
            take
            args << parse(0)
          end
        end
        expect(close)
        args
      end
    end
  end
end
